src/stackdriver-nozzle/nozzle/nozzle.go

```diff
@@ -22,45 +23,29 @@ func (e *PostContainerMetricError) Error() string {
 
 type Nozzle struct {
 	StackdriverClient stackdriver.Client
+	Serializer        serializer.Serializer
 }
 
-func (n *Nozzle) HandleEvent(eventsEnvelope *events.Envelope) error {
-	envelope := Envelope{eventsEnvelope}
-	labels := envelope.Labels()
-
-	switch envelope.GetEventType() {
-	case events.Envelope_ContainerMetric:
-		return n.postContainerMetrics(envelope)
-	case events.Envelope_ValueMetric:
-		valueMetric := envelope.GetValueMetric()
-		name := valueMetric.GetName()
-		value := valueMetric.GetValue()
-
-		err := n.StackdriverClient.PostMetric(name, value, labels)
-		return err
-	default:
-		n.StackdriverClient.PostLog(envelope, labels)
+func (n *Nozzle) HandleEvent(envelope *events.Envelope) error {
+	if n.Serializer.IsLog(envelope) {
+		log := n.Serializer.GetLog(envelope)
+		n.StackdriverClient.PostLog(log.Payload, log.Labels)
 		return nil
+	} else {
+		metrics := n.Serializer.GetMetrics(envelope)
```

Comment:
Firehose expects us to consume the events very, very fast. The serializer is now making calls back to CloudFoundry to build that App metadata. So we have an issue where if there is a series of cache misses, the firehose will drop our connection because we're not pulling events fast enough. So now both the serializer and the `stackdriver.client` can be slow enough for firehose to drop us.

Given we have a worker pool story, I don't know that we need to do anything for this PR at all - I just wanted to bring it up and make sure we circle back in that story to *both* potential slow spots.